RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc

```diff
@@ -419,40 +447,67 @@ DAClusterizerInZ_vect::split(const double beta,  track_t &tks, vertex_t & y ) co
 
   for(unsigned int ic=0; ic<critical.size(); ic++){
     unsigned int k=critical[ic].second;
+
     // estimate subcluster positions and weight
     double p1=0, z1=0, w1=0;
     double p2=0, z2=0, w2=0;
     for(unsigned int i=0; i<nt; i++){
-      if (tks._Z_sum[i] > 0) {
-	double p = y._pk[k] * local_exp(-beta * Eik(tks._z[i], y._z[k], tks._dz2[i])) / tks._Z_sum[i];
-        double w=p*tks._dz2[i];
-        if(tks._z[i]<y._z[k]){
-          p1+=p; z1+=w*tks._z[i]; w1+=w;
-        }else{
-          p2+=p; z2+=w*tks._z[i]; w2+=w;
-        }
+      if (tks._Z_sum[i] > 1.e-100) {
+
+	// winner-takes-all, usually overestimates splitting
+	double tl = tks._z[i] < y._z[k] ? 1.: 0.;
+	double tr = 1. - tl;
+
+	 // soften it, especially at low T
+	double arg = (tks._z[i] - y._z[k]) * sqrt(beta * tks._dz2[i]);
+	if(std::fabs(arg) < 20){
+	  double t = local_exp(-arg);
+	  tl = t/(t+1.);
+	  tr = 1/(t+1.);
+	}
+
+	double p = y._pk[k] * tks._pi[i] * local_exp(-beta * Eik(tks._z[i], y._z[k], tks._dz2[i])) / tks._Z_sum[i];
+	double w = p*tks._dz2[i];
+	p1 += p*tl ; z1 += w*tl*tks._z[i]; w1 += w*tl;
+	p2 += p*tr;  z2 += w*tr*tks._z[i]; w2 += w*tr;
       }
     }
-    if(w1>0){  z1=z1/w1;} else{z1=y._z[k]-epsilon;}
-    if(w2>0){  z2=z2/w2;} else{z2=y._z[k]+epsilon;}
+
+    if(w1>0){z1 = z1/w1;} else {z1=y._z[k]-epsilon;}
+    if(w2>0){z2 = z2/w2;} else {z2=y._z[k]+epsilon;}
 
     // reduce split size if there is not enough room
-    if( ( k   > 0 ) && ( y._z[k-1]>=z1 ) ){ z1=0.5*(y._z[k]+y._z[k-1]); }
-    if( ( k+1 < nv) && ( y._z[k+1]<=z2 ) ){ z2=0.5*(y._z[k]+y._z[k+1]); }
+    if( ( k   > 0 ) && ( z1 < (0.6*y._z[k] + 0.4*y._z[k-1])) ){ z1 = 0.6*y._z[k] + 0.4*y._z[k-1]; }
+    if( ( k+1 < nv) && ( z2 > (0.6*y._z[k] + 0.4*y._z[k+1])) ){ z2 = 0.6*y._z[k] + 0.4*y._z[k+1]; }
+
+    if(verbose_){
+      if (std::fabs(y._z[k] - zdumpcenter_) < zdumpwidth_){
+	std::cout << " T= " << std::setw(8) << 1./beta 
+		  << " Tc= " << critical[ic].first 
+		  << "    splitting " << std::fixed << std::setprecision(4) << y._z[k] 
+		  << " --> " << z1 << "," << z2 
+		  << "     [" << p1 << "," << p2 << "]" ;
+	if (std::fabs(z2-z1)>epsilon){
+	  std::cout << std::endl;
+	}else{
+	  std::cout <<  "  rejected " << std::endl;
+	}
+      }
+    }
 
     // split if the new subclusters are significantly separated
-    if( (z2-z1)>epsilon){
-      split=true;
-      double pk1=p1*y._pk[k]/(p1+p2);
-      double pk2=p2*y._pk[k]/(p1+p2);
+    if( (z2-z1) > epsilon){
```

Comment:
@werdmann 
while reviewing code updates in the ZT clustering https://github.com/cms-sw/cmssw/pull/19935#pullrequestreview-61011007
this was suggested to be a bug with a missing std::abs.
Please check and if is indeed a bug, please submit a PR with a fix.
Thank you.

Reply:
z2 > z1 by construction. Basically z1 is made from all tracks left of the old z and z2 from all tracks right of the old z. An additional abs should not make any difference.